Chi-square test: It is basically applied between two categorical variables from a single population to understand whether there is a signification association between the two categories.

Taking typical loan dataset,in this notebook let us try to find out whether there is any variation between Male and female based on Self_Employed categories.

Let us assume that null hypothesis as "There is no variation"
Alternate hypothesis as "There is a variation"

In [6]:
import pandas as pd
import numpy as np
df_loan=pd.read_csv('loan.csv')
df_loan.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

In [7]:
df_loan.columns

Index(['Loan_ID', 'Gender', 'Married', 'Dependents', 'Education',
       'Self_Employed', 'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Property_Area', 'Loan_Status'],
      dtype='object')

In [11]:
#inport the statistics library
import scipy as stats


pd.crosstab(): This function is used to compute a simple cross-tabulation of two (or more) factors. It takes the specified factors (in this case, 'Gender' and 'Self_Employed') and computes a table of frequencies. The result is a DataFrame.

In [12]:
df_loan_table=pd.crosstab(df_loan['Gender'],df_loan['Self_Employed'])
df_loan_table

Self_Employed   No  Yes
Gender                 
Female          89   15
Male           402   63

In [19]:
observed_values=df_loan_table.values
observed_values

array([[ 89,  15],
       [402,  63]], dtype=int64)

In [16]:
#perform a chi-square test for independence on the contingency table (df_loan_table) 


In [17]:
from scipy.stats import chi2_contingency

# Assuming df_loan_table is your contingency table
val = chi2_contingency(df_loan_table)

# Print the results
print("Chi-square Statistic:", val[0])
print("P-value:", val[1])
print("Degrees of Freedom:", val[2])
print("Expected Frequencies Table:")
Expected_Values=val[3]
print(val[3])



Chi-square Statistic: 0.005893176199177382
P-value: 0.9388088831667365
Degrees of Freedom: 1
Expected Frequencies Table:
[[ 89.74340949  14.25659051]
 [401.25659051  63.74340949]]


In [20]:
from scipy.stats import chi2
chi_square=sum([(o-e)**2./e for o,e in zip(observed_values,Expected_Values)])
chi_square_statistic=chi_square[0]+chi_square[1]

In [21]:
print("chi-square statistic:-",chi_square_statistic)

chi-square statistic:- 0.054970617322921536


In [24]:
alpha = 0.05
critical_value=chi2.ppf(q=1-alpha,df=val[2])
print('critical_value:',critical_value)

critical_value: 3.841458820694124


In [27]:
#p-value
p_value=1-chi2.cdf(x=chi_square_statistic,df=val[2])
print('p-value:',p_value)
print('Significance level: ',alpha)
print('Degree of Freedom: ',val[2])
print('p-value:',p_value)

p-value: 0.8146293304419566
Significance level:  0.05
Degree of Freedom:  1
p-value: 0.8146293304419566


In [28]:
if chi_square_statistic>=critical_value:
    print("Reject H0,There is a relationship between 2 categorical variables")
else:
    print("Retain H0,There is no relationship between 2 categorical variables")
    
if p_value<=alpha:
    print("Reject H0,There is a relationship between 2 categorical variables")
else:
    print("Retain H0,There is no relationship between 2 categorical variables")

Retain H0,There is no relationship between 2 categorical variables
Retain H0,There is no relationship between 2 categorical variables


Conclusions summary:In hypothesis testing, the conclusion is based on comparing the p-value to the significance level (α). The null hypothesis (H0) is rejected if the p-value is less than or equal to the significance level. The decision is made as follows:

If p-value ≤ α: Reject the null hypothesis (H0).
If p-value > α: Do not reject the null hypothesis (H0).
In our example:

Given Information:

Chi-square Statistic: 0.005893176199177382
P-value: 0.9388088831667365
Significance Level (α): 0.05
Degrees of Freedom: 1
Decision Rule:

If p-value ≤ 0.05, reject H0.
If p-value > 0.05, do not reject H0.
Conclusion:

P-value (0.9388) > Significance Level (0.05)
Since the p-value is greater than the significance level, you would not reject the null hypothesis. The conclusion is that there is not enough evidence to suggest a significant relationship between the variables. In other words, based on the observed data, you do not have sufficient evidence to claim a variation or association between the variables.

Please note that a higher p-value suggests weaker evidence against the null hypothesis, and in this case, the p-value is substantially higher than the significance level, supporting the decision not to reject the null hypothesis.